<a href="https://colab.research.google.com/github/luis-telesforo/Cleaning-Data/blob/main/managing_several_datasets_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Managing several datasets I

In [1]:
!pip install xlrd==1.2.0
import pandas as pd
import numpy as np

from google.colab import drive
drive.mount('/content/drive')

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 103 kB 5.5 MB/s 
  Attempting uninstall: xlrd
    Found existing installation: xlrd 1.1.0
    Uninstalling xlrd-1.1.0:
      Successfully uninstalled xlrd-1.1.0
Mounted at /content/drive


## The first dataset
We have a list of indicators of energy supply and renewable electricity production from the [United Nations](http://unstats.un.org/unsd/environment/excel_file_tables/2013/Energy%20Indicators.xls) for the year 2013.


In [2]:
#This time we have a xls file
Energy = pd.read_excel('/content/drive/MyDrive/bases de datos michigan/Energy Indicators.xls')
Energy

Unnamed: 0                                         Unnamed: 1  \
0           NaN                                                NaN   
1           NaN                                                NaN   
2           NaN                                                NaN   
3           NaN                                                NaN   
4           NaN                                                NaN   
..          ...                                                ...   
277         NaN                       … denotes no data available.   
278         NaN                                                NaN   
279         NaN                                     Data Quality:    
280         NaN  The data are compiled primarily from the annua...   
281         NaN  For more information, visit http://unstats.un....   

                                            Unnamed: 2 Unnamed: 3 Unnamed: 4  \
0                     Environmental Indicators: Energy        NaN        NaN   
1                                                  NaN        NaN        NaN   
2    Energy Supply and Renewable Electricity Produc...        NaN        NaN   
3                                                  NaN        NaN        NaN   
4                                                  NaN        NaN        NaN   
..                                                 ...        ...        ...   
277                                                NaN        NaN        NaN   
278                                                NaN        NaN        NaN   
279                                                NaN        NaN        NaN   
280                                                NaN        NaN        NaN   
281                                                NaN        NaN        NaN   

                     Unnamed: 5  
0                           NaN  
1                           NaN  
2                           NaN  
3                           NaN  
4    Last update: December 2015  
..                          ...  
277                         NaN  
278                         NaN  
279                         NaN  
280                         NaN  
281                         NaN  

[282 rows x 6 columns]

Our xls file has a format that looks great but we do not need it to look great. We drop its header and footer; also we fix column names. Observe that row 277 is an annotation: the string `...` denotes 'no data available'. So, we need to fix that too.

In [3]:
#We drop footer and header
Energy = pd.read_excel('/content/drive/MyDrive/bases de datos michigan/Energy Indicators.xls',skiprows=16,nrows=228)

#First two columns and the first row contain no essential data
Energy.drop(columns=['Unnamed: 0','Unnamed: 1'],index=0,inplace=True)

#We add adequate labels
Energy.rename({'Unnamed: 2':'Country','Energy Supply per capita':'Energy Supply per Capita','Renewable Electricity Production':'% Renewable'},axis=1,inplace=True)

#Some country names do not coincide with the other data set
#we homogenize them
Energy.replace(to_replace=['(\(.*\))','(\d*)'],value=['',''],regex=True,inplace=True)
Energy.replace(to_replace=['Republic of Korea','United States of America','United Kingdom of Great Britain and Northern Ireland','China, Hong Kong Special Administrative Region'],value=['South Korea','United States','United Kingdom','Hong Kong'],inplace=True)
Energy['Country']=Energy['Country'].str.strip()
#Handling missing values
Energy.replace(to_replace='...',value=np.nan,inplace=True)

#Finally, we need the same units of energy
Energy['Energy Supply'] = Energy['Energy Supply'].apply(lambda x: x*1000000)

Energy

Country  Energy Supply  Energy Supply per Capita  \
1                  Afghanistan   3.210000e+08                      10.0   
2                      Albania   1.020000e+08                      35.0   
3                      Algeria   1.959000e+09                      51.0   
4               American Samoa            NaN                       NaN   
5                      Andorra   9.000000e+06                     121.0   
..                         ...            ...                       ...   
223                   Viet Nam   2.554000e+09                      28.0   
224  Wallis and Futuna Islands   0.000000e+00                      26.0   
225                      Yemen   3.440000e+08                      13.0   
226                     Zambia   4.000000e+08                      26.0   
227                   Zimbabwe   4.800000e+08                      32.0   

     % Renewable  
1      78.669280  
2     100.000000  
3       0.551010  
4       0.641026  
5      88.695650  
..           ...  
223    45.321520  
224     0.000000  
225     0.000000  
226    99.714670  
227    52.536120  

[227 rows x 4 columns]

##The Second dataset

We have a csv containing countries' GDP from 1960 to 2015 from [World Bank](http://data.worldbank.org/indicator/NY.GDP.MKTP.CD). (This dataset is updated continuously)

First 4 rows are empty.

In [4]:
GDP = pd.read_csv('/content/drive/MyDrive/bases de datos michigan/world_bank.csv',skiprows=4)
GDP

Country Name Country Code                            Indicator Name  \
0               Aruba          ABW  GDP at market prices (constant 2010 US$)   
1             Andorra          AND  GDP at market prices (constant 2010 US$)   
2         Afghanistan          AFG  GDP at market prices (constant 2010 US$)   
3              Angola          AGO  GDP at market prices (constant 2010 US$)   
4             Albania          ALB  GDP at market prices (constant 2010 US$)   
..                ...          ...                                       ...   
259       Yemen, Rep.          YEM  GDP at market prices (constant 2010 US$)   
260      South Africa          ZAF  GDP at market prices (constant 2010 US$)   
261  Congo, Dem. Rep.          COD  GDP at market prices (constant 2010 US$)   
262            Zambia          ZMB  GDP at market prices (constant 2010 US$)   
263          Zimbabwe          ZWE  GDP at market prices (constant 2010 US$)   

     Indicator Code          1960          1961          1962          1963  \
0    NY.GDP.MKTP.KD           NaN           NaN           NaN           NaN   
1    NY.GDP.MKTP.KD           NaN           NaN           NaN           NaN   
2    NY.GDP.MKTP.KD           NaN           NaN           NaN           NaN   
3    NY.GDP.MKTP.KD           NaN           NaN           NaN           NaN   
4    NY.GDP.MKTP.KD           NaN           NaN           NaN           NaN   
..              ...           ...           ...           ...           ...   
259  NY.GDP.MKTP.KD           NaN           NaN           NaN           NaN   
260  NY.GDP.MKTP.KD  7.707081e+10  8.003449e+10  8.497844e+10  9.124438e+10   
261  NY.GDP.MKTP.KD  1.508024e+10  1.344383e+10  1.629401e+10  1.714358e+10   
262  NY.GDP.MKTP.KD  4.592975e+09  4.655503e+09  4.539542e+09  4.688093e+09   
263  NY.GDP.MKTP.KD  3.338344e+09  3.549199e+09  3.600111e+09  3.824915e+09   

             1964          1965  ...          2006          2007  \
0             NaN           NaN  ...           NaN           NaN   
1             NaN           NaN  ...  4.018196e+09  4.021331e+09   
2             NaN           NaN  ...  1.030523e+10  1.172119e+10   
3             NaN           NaN  ...  5.581103e+10  6.842044e+10   
4             NaN           NaN  ...  9.771760e+09  1.034829e+10   
..            ...           ...  ...           ...           ...   
259           NaN           NaN  ...  2.672565e+10  2.761787e+10   
260  9.848904e+10  1.072507e+11  ...  3.402852e+11  3.585261e+11   
261  1.672524e+10  1.689205e+10  ...  1.650894e+10  1.754232e+10   
262  5.260699e+09  6.136472e+09  ...  1.440569e+10  1.560892e+10   
263  3.782605e+09  3.968352e+09  ...  1.006276e+10  9.695130e+09   

             2008          2009          2010          2011          2012  \
0             NaN           NaN  2.467704e+09           NaN           NaN   
1    3.675728e+09  3.535389e+09  3.346317e+09  3.185605e+09  3.129538e+09   
2    1.214448e+10  1.469733e+10  1.593680e+10  1.691113e+10  1.935220e+10   
3    7.787420e+10  7.975320e+10  8.247091e+10  8.570262e+10  9.012096e+10   
4    1.112752e+10  1.150029e+10  1.192695e+10  1.223109e+10  1.240477e+10   
..            ...           ...           ...           ...           ...   
259  2.872656e+10  2.991436e+10  3.090675e+10  2.624342e+10  2.689160e+10   
260  3.699668e+11  3.642764e+11  3.753494e+11  3.874074e+11  3.960071e+11   
261  1.863448e+10  1.916651e+10  2.052329e+10  2.193213e+10  2.350200e+10   
262  1.682234e+10  1.837342e+10  2.026555e+10  2.140358e+10  2.302438e+10   
263  7.982103e+09  8.459783e+09  9.422161e+09  1.054391e+10  1.165789e+10   

             2013          2014          2015  
0             NaN           NaN           NaN  
1    3.127550e+09           NaN           NaN  
2    1.973134e+10  1.999032e+10  2.029415e+10  
3    9.626143e+10  1.008863e+11  1.039106e+11  
4    1.254247e+10  1.279331e+10  1.312082e+10  
..            ...           ...           ...  
259  2.800914e+10    

We homogenize country names. Also, we only work with 2006-2015 data.

In [5]:
GDP.replace({"Korea, Rep.": "South Korea", "Iran, Islamic Rep.": "Iran", "Hong Kong SAR, China": "Hong Kong"},inplace=True)
GDP.rename({'Country Name':'Country'},axis=1,inplace=True)
GDP['Country']=GDP['Country'].str.strip()

GDP.drop(GDP.columns[1:50],inplace=True,axis=1)
GDP

Country          2006          2007          2008          2009  \
0               Aruba           NaN           NaN           NaN           NaN   
1             Andorra  4.018196e+09  4.021331e+09  3.675728e+09  3.535389e+09   
2         Afghanistan  1.030523e+10  1.172119e+10  1.214448e+10  1.469733e+10   
3              Angola  5.581103e+10  6.842044e+10  7.787420e+10  7.975320e+10   
4             Albania  9.771760e+09  1.034829e+10  1.112752e+10  1.150029e+10   
..                ...           ...           ...           ...           ...   
259       Yemen, Rep.  2.672565e+10  2.761787e+10  2.872656e+10  2.991436e+10   
260      South Africa  3.402852e+11  3.585261e+11  3.699668e+11  3.642764e+11   
261  Congo, Dem. Rep.  1.650894e+10  1.754232e+10  1.863448e+10  1.916651e+10   
262            Zambia  1.440569e+10  1.560892e+10  1.682234e+10  1.837342e+10   
263          Zimbabwe  1.006276e+10  9.695130e+09  7.982103e+09  8.459783e+09   

             2010          2011          2012          2013          2014  \
0    2.467704e+09           NaN           NaN           NaN           NaN   
1    3.346317e+09  3.185605e+09  3.129538e+09  3.127550e+09           NaN   
2    1.593680e+10  1.691113e+10  1.935220e+10  1.973134e+10  1.999032e+10   
3    8.247091e+10  8.570262e+10  9.012096e+10  9.626143e+10  1.008863e+11   
4    1.192695e+10  1.223109e+10  1.240477e+10  1.254247e+10  1.279331e+10   
..            ...           ...           ...           ...           ...   
259  3.090675e+10  2.624342e+10  2.689160e+10  2.800914e+10           NaN   
260  3.753494e+11  3.874074e+11  3.960071e+11  4.047682e+11  4.110369e+11   
261  2.052329e+10  2.193213e+10  2.350200e+10  2.550050e+10  2.778776e+10   
262  2.026555e+10  2.140358e+10  2.302438e+10  2.420595e+10  2.542227e+10   
263  9.422161e+09  1.054391e+10  1.165789e+10  1.218064e+10  1.264939e+10   

             2015  
0             NaN  
1             NaN  
2    2.029415e+10  
3    1.039106e+11  
4    1.312082e+10  
..            ...  
259           NaN  
260  4.163117e+11  
261  2.970961e+10  
262  2.624127e+10  
263  1.278517e+10  

[264 rows x 11 columns]

##The third dataset

Finally, we took the [Sciamgo Journal and Country Rank data for Energy Engineering and Power Technology](http://www.scimagojr.com/countryrank.php?category=2102) which ranks countries based on their journal contributions in the aforementioned area (the dataset corrersponds to 1996-2013)

In [6]:
ScimEn = pd.read_excel('/content/drive/MyDrive/bases de datos michigan/scimagojr-3.xlsx')
ScimEn['Country']=ScimEn['Country'].str.strip()
ScimEn

Rank             Country  Documents  Citable documents  Citations  \
0       1               China     127050             126767     597237   
1       2       United States      96661              94747     792274   
2       3               Japan      30504              30287     223024   
3       4      United Kingdom      20944              20357     206091   
4       5  Russian Federation      18534              18301      34266   
..    ...                 ...        ...                ...        ...   
186   187              Guyana          1                  1          0   
187   188    Christmas Island          1                  1          0   
188   189             Reunion          1                  1          2   
189   190         Saint Lucia          1                  1          0   
190   191          Mauritania          1                  1          1   

     Self-citations  Citations per document  H index  
0            411683                    4.70      138  
1            265436                    8.20      230  
2             61554                    7.31      134  
3             37874                    9.84      139  
4             12422                    1.85       57  
..              ...                     ...      ...  
186               0                    0.00        0  
187               0                    0.00        0  
188               1                    2.00        1  
189               0                    0.00        0  
190               0                    1.00        1  

[191 rows x 8 columns]

#Merging our datasets
Now, we join the three datasets: GDP, Energy, and ScimEn into a new dataset (using the intersection of country names) and we take the top 15 countries by Scimagojr 'Rank' only. 

In [7]:
def top15():
  df = pd.merge(ScimEn,Energy,how='inner')
  df1 =pd.merge(df,GDP,how='inner')
  #we drop ranks >15
  df1.drop(df1.index[15:],inplace=True)
  df1.set_index('Country',inplace=True)
  return df1

top15()

Rank  Documents  Citable documents  Citations  \
Country                                                             
China                  1     127050             126767     597237   
United States          2      96661              94747     792274   
Japan                  3      30504              30287     223024   
United Kingdom         4      20944              20357     206091   
Russian Federation     5      18534              18301      34266   
Canada                 6      17899              17620     215003   
Germany                7      17027              16831     140566   
India                  8      15005              14841     128763   
France                 9      13153              12973     130632   
South Korea           10      11983              11923     114675   
Italy                 11      10964              10794     111850   
Spain                 12       9428               9330     123336   
Iran                  13       8896               8819      57470   
Australia             14       8831               8725      90765   
Brazil                15       8668               8596      60702   

                    Self-citations  Citations per document  H index  \
Country                                                               
China                       411683                    4.70      138   
United States               265436                    8.20      230   
Japan                        61554                    7.31      134   
United Kingdom               37874                    9.84      139   
Russian Federation           12422                    1.85       57   
Canada                       40930                   12.01      149   
Germany                      27426                    8.26      126   
India                        37209                    8.58      115   
France                       28601                    9.93      114   
South Korea                  22595                    9.57      104   
Italy                        26661                   10.20      106   
Spain                        23964                   13.08      115   
Iran                         19125                    6.46       72   
Australia                    15606                   10.28      107   
Brazil                       14396                    7.00       86   

                    Energy Supply  Energy Supply per Capita  % Renewable  \
Country                                                                    
China                1.271910e+11                      93.0    19.754910   
United States        9.083800e+10                     286.0    11.570980   
Japan                1.898400e+10                     149.0    10.232820   
United Kingdom       7.920000e+09                     124.0    10.600470   
Russian Federation   3.070900e+10                     214.0    17.288680   
Canada               1.043100e+10                     296.0    61.945430   
Germany              1.326100e+10                     165.0    17.901530   
India                3.319500e+10                      26.0    14.969080   
France               1.059700e+10                     166.0    17.020280   
South Korea          1.100700e+10                     221.0     2.279353   
Italy                6.530000e+09                     109.0    33.667230   
Spain                4.923000e+09                     106.0    37.968590   
Iran                 9.172000e+09                     119.0     5.707721   
Australia            5.386000e+09                     231.0    11.810810   
Brazil               1.214900e+10                      59.0    69.648030   

                            2006          2007          2008          2009  \
Country                                                                      
China               3.992331e+12  4.559041e+12  4.997775e+12  5.459247e+12   
United States       1.479230e+13  1.505540e+13  1.501149e+13  1.459484e+13   
Japan               5.496542e+12 

What are the top 15 countries for average GDP over the last 10 years?

In [9]:
def gdp_average():
  gdp = ['2006','2007','2008','2009','2010','2011','2012','2013','2014','2015']
  return top15()[gdp].apply(np.nanmean,axis=1).rename('avgGDP').sort_values(ascending=False)
    # YOUR CODE HERE
gdp_average()

Country
United States         1.536434e+13
China                 6.348609e+12
Japan                 5.542208e+12
Germany               3.493025e+12
France                2.681725e+12
United Kingdom        2.487907e+12
Brazil                2.189794e+12
Italy                 2.120175e+12
India                 1.769297e+12
Canada                1.660647e+12
Russian Federation    1.565459e+12
Spain                 1.418078e+12
Australia             1.164043e+12
South Korea           1.106715e+12
Iran                  4.441558e+11
Name: avgGDP, dtype: float64

We estimate the population of each country in **top15** using *Energy Supply* and *Energy Supply per capita* columns and applying a funtion to **top15**

In [27]:
def pop(row):
  return row['Energy Supply']/row['Energy Supply per Capita']
#we will need pop_est to be a data frame, so, we do not
#use apply method directly
est_pop = pd.DataFrame(top15()[['Energy Supply','Energy Supply per Capita']].apply(pop,axis=1).rename('Population',axis=1))
est_pop

Population
Country                         
China               1.367645e+09
United States       3.176154e+08
Japan               1.274094e+08
United Kingdom      6.387097e+07
Russian Federation  1.435000e+08
Canada              3.523986e+07
Germany             8.036970e+07
India               1.276731e+09
France              6.383735e+07
South Korea         4.980543e+07
Italy               5.990826e+07
Spain               4.644340e+07
Iran                7.707563e+07
Australia           2.331602e+07
Brazil              2.059153e+08

What is the correlation between the number of citable documents per capita and the energy supply per capita? 

In [28]:
#We add to pop_est the number of citable documents
est_pop['Citable documents'] = top15()['Citable documents']

#with the following we calculate the number of citable documents per capita
def doc_per(row):
  return row['Citable documents']/row['Population']

#Finally, we calculate the correlation.
est_pop.apply(doc_per,axis=1).corr(top15()['Energy Supply per Capita'])

0.7940010435442946

Using the following dictionary we group the Countries by Continent, then create a DataFrame that displays the sample size (the number of countries in each continent bin), and the sum, mean, and std deviation for the estimated population of each country.

```python
ContinentDict  = {'China':'Asia', 
                  'United States':'North America', 
                  'Japan':'Asia', 
                  'United Kingdom':'Europe', 
                  'Russian Federation':'Europe', 
                  'Canada':'North America', 
                  'Germany':'Europe', 
                  'India':'Asia',
                  'France':'Europe', 
                  'South Korea':'Asia', 
                  'Italy':'Europe', 
                  'Spain':'Europe', 
                  'Iran':'Asia',
                  'Australia':'Australia', 
                  'Brazil':'South America'}
```

In [32]:
ContinentDict  = {'China':'Asia','United States':'North America', 'Japan':'Asia', 'United Kingdom':'Europe', 'Russian Federation':'Europe', 'Canada':'North America', 'Germany':'Europe', 'India':'Asia', 'France':'Europe', 'South Korea':'Asia', 'Italy':'Europe', 'Spain':'Europe', 'Iran':'Asia', 'Australia':'Australia', 'Brazil':'South America'}

#we add new columns to est_pop, on them we calculate the sum, mean...
est_pop['size'] = est_pop['Population']
est_pop['sum'] = est_pop['Population']
est_pop['mean'] = est_pop['Population']
est_pop['std'] = est_pop['Population']

#we extract the continent of each country
def grp_det(item):
  return ContinentDict[item]
#and we use groupby with the dictionary
est_pop.groupby(by=grp_det).aggregate({'size':len,'sum':np.sum,'mean':np.mean,'std':np.std})


size           sum          mean           std
Country                                                      
Asia              5  2.898666e+09  5.797333e+08  6.790979e+08
Australia         1  2.331602e+07  2.331602e+07           NaN
Europe            6  4.579297e+08  7.632161e+07  3.464767e+07
North America     2  3.528552e+08  1.764276e+08  1.996696e+08
South America     1  2.059153e+08  2.059153e+08           NaN